### ISSUE TITLE
Integration test for spare-kafka node failure

### ISSUE THREAD

**Issue by KingOfPoptart** [DemandCube/NeverwinterDP_Legacy#679]

**Comment by KingOfPoptart** [comment]
Requirements:
1. Create a test using the kafka stability check tool
2. Set the test up in Jenkins - Set it up to run once a week

Development plan:
1. Create a first test that kills a single kafka node and restarts it on a spare node.  Show that no messages are lost.  Share the results with the team for review
2. Expand the test to show it can handle 2 failures with 2 spare kafka nodes
3. Expand the test to run continuously for 1,000,000 messages

**Comment by jeroldleslie** [comment]
Integration test senario for kafka failure simulation using spare nodes
1. single node failure
   1. this should kill/start spare nodes which selected randomly from the kafka cluster (only one node at a time). If new node added reassigning replicas will be done. this process will be looped for the lifetime. 
   2. kafka stability check tool is used to push and retrieve messages into the kafka cluster.
   3. finally the test should pass with the given maximum messages (up to 10 million messages) without loss. (this is the expecting result)
2. multi node failure
   1. this should kill/start spare nodes which selected randomly from the kafka cluster (one node or more than one at a time). If new node added reassigning replicas will be done. this process will be looped for the lifetime. 
   2. kafka stability check tool is used to push and retrieve messages into the kafka cluster.
   3. finally the test should pass with the given maximum messages (up to 10 million messages) without loss. (this is the expecting result)

Adding more functionality
To make the test more specific for spare node adding restart-method command with options  “random” and flipflop. (also this options applicable for single node failure)
        flipflop
            kill/shutdown random nodes either be running normal-kafka or running  spare-kaka
            restart nodes random died normal-kaka if spare-kaka died and died spare-kafka if normal-kaka died
        random
            kill/shutdown random nodes with combination of both running normal-kaka and running spare-kaka
            restart random nodes with combination of both died normal-kaka and died spare-kaka

**Comment by tuan08** [comment]
I think the code and concept is not well designed from the begining for this type of feature. We just think to have the spare node later and replace the dead node with the spare node. 

A better code organization would be:
1. Allocate a set node for  for server type , for ex 5 nodes for kafka
2. When you launch , you launch 3 kafka node only and mark the server as kafka role, mark the 2 other as spare-kafka
3. When you kill and replace:
4.  select randomly one, kill and clean all its data, configuration ... 
5. select the spare one and launch 
6. move the killed one to spare

**Comment by tuan08** [comment]
You should not have the concept of single node or multiple node, as you select the node randomly.  You should have the concept how many time do you want to kill or replace.

**Comment by jeroldleslie** [comment]
yes.. all the 6 points are done here. having kafka role for normal nodes and spare-kafka for spare nodes. and single node and multiple node is not concept it is test plan. there in failure we dont have single node failure or multiple node failure. ther is an option --servers-to-fail-simultaneously which says number of nodes to kill simultanously. if you give 1 it work as a single node failure if you give value greater than 1 it will pick random number of nodes to kill/shutdown

**Comment by tuan08** [comment]
I remember that we have the host name as kafka and spare-kafka which is not work the way I describe

**Comment by tuan08** [comment]
Also , just need to select randomly one kill and replace , no need to work on a specific node

**Comment by jeroldleslie** [comment]
yes. not working on specific nodes. just doing it randomly

**Comment by tuan08** [comment]
Since your cluster commander can be run several times , so you may need to parse the host file  , check the running process , server name  to mark the working , and spare server correctly. We rely on the host name to identify the working and spare , which is not good and dynamic enough
